Wikipedia/Code/SaveButtonsController.swift

```diff
@@ -0,0 +1,63 @@
+import UIKit
+
+@objc(WMFSaveButtonsController) class SaveButtonsController: NSObject {
+    
+    var visibleSaveButtons = [Int: SaveButton]()
+    var visibleArticleKeys = [Int: String]()
+    let dataStore: MWKDataStore
+    let savedPagesFunnel = SavedPagesFunnel()
+    
+    required init(dataStore: MWKDataStore) {
+        self.dataStore = dataStore
+        super.init()
+        NotificationCenter.default.addObserver(self, selector: #selector(articleUpdated(notification:)), name: NSNotification.Name.WMFArticleUpdated, object: nil)
+    }
+    
+    deinit {
+        NotificationCenter.default.removeObserver(self)
+    }
+    
+    @objc(willDisplaySaveButton:forArticle:) func willDisplay(saveButton: SaveButton, for article: WMFArticle) {
```

Comment:
Yeah, I like for class names too. The inline `@objc` isn't as bad if it doesn't have parameters, but I still prefer 2 lines. Maybe it's just because I'm used to other languages that use that formatting.

http://www.saltycrane.com/blog/2010/03/simple-python-decorator-examples/
https://docs.oracle.com/javase/tutorial/java/annotations/basics.html